Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -963,24 +964,57 @@ extension SourceKitServer: BuildSystemDelegate {
   }
 }
 
+extension LanguageServerProtocol.BuildConfiguration {
+  /// Convert `LanguageServerProtocol.BuildConfiguration` to `SKSupport.BuildConfiguration`.
+  var configuration: SKSupport.BuildConfiguration {
+    switch self {
+    case .debug: return .debug
+    case .release: return .release
+    }
+  }
+}
+
 // MARK: - Request and notification handling
 
 extension SourceKitServer {
 
   // MARK: - General
 
+  /// Returns the build setup for the parameters specified for the given `WorkspaceFolder`.
+  private func buildSetup(for workspaceFolder: WorkspaceFolder) -> BuildSetup {
+    let buildParams = workspaceFolder.buildSetup
+    let scratchPath: AbsolutePath?
+    if let scratchPathParam = buildParams?.scratchPath {
+      scratchPath = try? AbsolutePath(validating: scratchPathParam.pseudoPath)
+    } else {
+      scratchPath = nil
+    }
+    return SKCore.BuildSetup(
+      configuration: workspaceFolder.buildSetup?.buildConfiguration?.configuration,
+      path: scratchPath,
+      flags: BuildFlags(
+        cCompilerFlags: buildParams?.cFlags ?? [],
+        cxxCompilerFlags: buildParams?.cxxFlags ?? [],
+        swiftCompilerFlags: buildParams?.swiftFlags ?? [],
+        linkerFlags: buildParams?.linkerFlags ?? [],
+        xcbuildFlags: []
+      )
+    )
+  }
+
   /// Creates a workspace at the given `uri`.
-  private func createWorkspace(uri: DocumentURI) async -> Workspace? {
+  private func createWorkspace(_ workspaceFolder: WorkspaceFolder) async -> Workspace? {
     guard let capabilityRegistry = capabilityRegistry else {
       logger.log("Cannot open workspace before server is initialized")
       return nil
     }
+    let workspaceBuildSetup = self.buildSetup(for: workspaceFolder)
     return try? await Workspace(
       documentManager: self.documentManager,
-      rootUri: uri,
+      rootUri: workspaceFolder.uri,
       capabilityRegistry: capabilityRegistry,
       toolchainRegistry: self.toolchainRegistry,
-      buildSetup: self.options.buildSetup,
+      buildSetup: workspaceBuildSetup.merging(self.options.buildSetup),
```

Comment:
Is it expected that the main options can override the workspace options? I would have expected it to be the other way around

Reply:
OK, I’ll flip the precedence order.